website/content/docs/secrets/gcp.mdx

```diff
@@ -480,6 +511,15 @@ bigquery.datasets.get
 bigquery.datasets.update
 ```
 
+When using service account impersonation, Vault must have the following additional
+permissions:
+
+```text
+# impersonation
+roles/iam.serviceAccountTokenCreator
+roles/iam.workloadIdentityUser
+```
+
```

Comment:
Not sure if I am providing this info correctly and if this is correct location to do so.

Reply:
The formatting and location seem fine to me. I can't speak for the correctness of the required permissions. Is that was you were referring to?

Reply:
I believe the permissions needed are `iam.serviceAccountTokenCreator` and `iam.serviceAccountKeyAdmin` (see [the acceptance tests configuration](https://github.com/hashicorp/vault-plugin-secrets-gcp/commit/18d1fa3756da73fac75ba48b1fd52ab6b1b238a3#diff-5ccaa7ae143e28068ecdb6a41a4526ede72645e725116decfb76e0b1952aad35R41-R58))

We _may_ be able to remove `iam.serviceAccountTokenCreator` from just looking at our IAM configuration that we use in our environments but I'm not positive about that and would need to test that to be sure.

Reply:
Ah right, I should probably explain where that second permission came from.

[This page](https://cloud.google.com/iam/docs/impersonating-service-accounts#allow-impersonation) under the section named "Workload Identity User", says that `roles/workloadIdentityUser` is needed to impersonate service accounts from GKE workloads. I've never personally used GKE so I am not sure if it's even relevant or usable through the GCP secret engine, but that's why I added it here first and then left a comment to follow up.

I will try to find out whether it has no use here, or perhaps you have experience with it Matt?

Reply:
I actually think this section you added here can be removed and instead the section on [static account permissions](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#LL472-L487) can be updated to read:

```text
When using static accounts or impersonated accounts, Vault must have the following permissions at the service account level:
```

A couple reasons for this:
1. All of the above are listing GCP permissions but what you've added are GCP roles which aggregate permissions and can be more permissive than desirable by users.
2. For impersonated accounts, this plugin only needs to be able to [get/lookup a service account](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/18d1fa3756da73fac75ba48b1fd52ab6b1b238a3/plugin/impersonated_account.go#L113). This is already covered [above in the required permissions](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L463).
3. The key permission to all of this is that the plugin's configured GCP Service Account has `iam.serviceAccounts.getAccessToken` permission **on the service account to impersonate**. This can be at the project, folder, or service account level but we should push users towards declaring it at the service account level which looks like is [already done above](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L479-L480).

The only other thing I would maybe do is remove [the following reference](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L475-L477) in hopes to reduce the likelihood of someone unknowingly adding `roles/iam.serviceAccountKeyAdmin` to the wrong service account:

```
# Service account key admin
# This is provided in the `roles/iam.serviceAccountKeyAdmin` role
iam.serviceAccounts.get
```

Reply:
Thanks for pointing that out, you're right the roles don't belong being listed there. I agree that simply repurposing the permissions of static accounts to include impersonated accounts is better and it makes it more concise. I think the note about serviceAccountKeyAdmin doesn't quite belong there either, plus it's already mentioned in the section below so I'll remove it.